[Shah Dhruv Nilesh] iP

dhshah1 opened the pull request:
## Duke - A task manager

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke helps you manage your tasks and increase your productivity. It's
* text based
* easy to use
* lightweight and ~~FAST~~ *SUPER* FAST
* Completely **FREE**

To get started:
1. Download the JAR file from [here](https://github.com/dhshah1/ip/releases/tag/A-Jar)
1. Double click the JAR file
1. Add your tasks
1. Watch your productivity increase :sunglasses:

Features
- [x] Manages Todos, Deadlines and Events
- [x] Ability to search by tasks by keywords
- [x] Store tasks on disk
- [ ] Reminders (coming soon)
- [ ] GUI (coming soon)

If you're a Java Programmer, feel free to check out the code. Here's part of the `run` method of the `Duke` class:

```java
while (!isExit) {
    try {
        String userInput = ui.getUserInput();
        Command userCommand = Parser.parse(userInput);
        userCommand.execute(tasks, ui);
        isExit = userCommand.isExit();
    } catch (UnsupportedCommandException | MalformedCommandException e) {
        ui.showErrorMessage(e);
    }
}
```

Commit:
Add Gradle support

dhshah1 committed:
Set up skeletal version of Duke

dhshah1 committed:
Add ability for chatbot to add and list tasks

dhshah1 committed:
Add ability for chatbot to mark task as done

dhshah1 committed:
Add support for different type of tasks

dhshah1 committed:
Add input and expeted txt files for testing

dhshah1 committed:
Add error handling for incorrect user input

dhshah1 committed:
Add ability to delete tasks

dhshah1 committed:
Add ability to store and load tasks from disk

dhshah1 committed:
Parse dates for deadline and events

dhshah1 committed:
Merge branch 'branch-Level-7'

dhshah1 committed:
Merge branch 'branch-Level-8'

dhshah1 committed:
Refactor code to group similar functionality into separate classes

dhshah1 committed:
Refactor classes into packages

dhshah1 committed:
Add some simple unit tests for tasks

dhshah1 committed:
Add some simple header comments to public methods

dhshah1 committed:
Fix some spacing coding violations

dhshah1 committed:
Add command to filter tasks by a keyword

dhshah1 committed:
Merge branch 'branch-A-CodingStandard'

dhshah1 committed:
Merge branch 'branch-Level-9'

dhshah1 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

cliveong commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+        try {
+            int taskIndex = Integer.parseInt(userInput.split(" ")[1]);
+            return new DeleteCommand(taskIndex);
+        } catch (ArrayIndexOutOfBoundsException | NumberFormatException  e) {
```
I think there is an extra spacing between NumberFormatException and e

cliveong commented on src/main/java/duke/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,33 @@
+            return new AddTodoCommand(task);
+        } catch (NullPointerException | ArrayIndexOutOfBoundsException e) {
+            throw new MalformedCommandException("Creating an todo needs to follow the following format: " +
+                "todo [description]");
```
The 8 spaces needed upon indentation seems to be missing

cliveong commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param description Task's description.
+     */
+    public Task(String description){
```
You might have missed out spacing between the constructor and the brackets

cliveong reviewed (commented):
Great work overall, was very difficult to spot any issues. Have much to learn from.

jivesh commented on src/main/java/duke/command/MalformedCommandException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Signals a user command that was incorrectly formatted.
+ */
+public class MalformedCommandException extends Exception {
```
Good name for the exception, clearly differentiates it from `UnsupportedCommandException`!

jivesh commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("d/MM/yyyy HHmm");
+    private static final DateTimeFormatter DISPLAY_FORMAT = DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a");
+
+    private LocalDateTime by;
```
While this is a private variable, maybe a more descriptive name would help. That being said, this short name is pretty elegant too, but personally I would optimise for readability and use a longer name.

jivesh commented:
Really good code overall! Lots of lessons here.

Zhouyirui commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+     * @param file File storing tasks on disk.
+     * @throws StorageException If file storing the tasks is incorrectly formatted.
+     */
+    public void loadTasks(File file) throws StorageException {
```
Perhaps storage should be in charge of reading the data from the file? I think it would be better if Storage passes an ArrayList of tasks for TaskList to initialise

Zhouyirui commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * Returns a Exit command.
+     *
+     * @return Xxit command.
```
Typo 😄

Zhouyirui commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public interface Command {
+    /**
+     *
```
Perhaps you can add short description of what execute function does?

Zhouyirui commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,40 @@
+     */
+    public static Command create(String userInput) throws MalformedCommandException {
+        try {
+            String userParams = userInput.split(" ", 2)[1];
+            String[] userParamsSplit = userParams.split(" /", 2);
+            String description = userParamsSplit[0];
```
You are using these code in AddEventCommand as well. Maybe you make it into one of Parser's functions?

Zhouyirui reviewed (commented):
Overall, neat code with well done OOP. Good job!

dhshah1 committed:
Fix minor spacing and indentation issues

dhshah1 commented:
@cliveong @jivesh @Zhouyirui Thanks for taking the time to review the code and suggesting improvements!

dhshah1 committed:
Create GUI for chatbot

dhshah1 committed:
Merge branch 'branch-Level-10'

dhshah1 committed:
Refactor code to improve code quality

dhshah1 committed:
Add some assertions to document important assumptions

dhshah1 committed:
Merge pull request #2 from dhshah1/branch-A-CodeQuality

Refactor Code to improve code quality

dhshah1 committed:
Merge branch 'master' into branch-A-Assertions

dhshah1 committed:
Merge pull request #3 from dhshah1/branch-A-Assertions

Add some assertions to the codebase

dhshah1 committed:
Add option to archive current list of tasks

There is no way to archive tasks.

This causes the list of tasks to continue to grow over time.
It is not user friendly to view a long list of tasks
in the GUI.

Let's provide a command for the user to specify the filename
to which the current tasks will be archived to.

dhshah1 committed:
Merge pull request #4 from dhshah1/branch-C-Archive

Add option to archive current list of tasks

dhshah1 committed:
Update README

dhshah1 committed:
Merge branch 'master' of https://github.com/dhshah1/ip

dhshah1 committed:
Set theme jekyll-theme-cayman
